Draft review for default extras PEP (#1)
astrofrog/peps: peps/pep-9999.rst

PEP: 9999
Title: Default Extras for Python Software Packages
Author: Thomas Robitaille <thomas.robitaille@gmail.com>,

>         Jonathan Dekhtiar <jonathan@dekhtiar.com>

Reviewer: ![image](https://github.com/user-attachments/assets/ec2cb2de-24c6-4115-b30c-3acb3535f0e0)
There's probably a need to adjust the formatting differently. I don't think that's how you intended it renders. Probably you have to put line 3 & 4 on the same line 

See how PEP 683 is doing it: https://raw.githubusercontent.com/astrofrog/peps/435dfce9e9f727c1458021e681f336021872f952/peps/pep-0683.rst

Author: Fixed (I'll push my changes once I have finished implementing all of them)

Sponsor: TBD
Status: Draft
Type: Standards Track
Topic: Packaging
Created: 24-Sep-2024

Abstract
========

:pep:`508` specifies a mini-language for
declaring package dependencies. One feature of this language is the ability to
specify *extras*, which are optional components of a distribution that, when
used, install additional dependencies. This PEP proposes a mechanism to allow
one or more extras to be installed by default if none are provided explicitly.

Motivation
==========

Various use cases for default extras and possible solutions in this PEP were discussed

> extensively at https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/38.

Reviewer 2: ```suggestion
extensively on `this DPO thread <https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/38>`__.
```

In this section we take a look at two common use cases that provide the
motivation for the present PEP.

Recommended but not required dependencies
-----------------------------------------

Package maintainers often use extras to declare optional dependencies that
extend the functionality or performance of a package. In some cases, it can be
difficult to determine which dependencies should be required and which should be
categorized as extras. A balance must be struck between the needs of typical
users (who may prefer most features to be available 'by default') and users who
want minimal installations without large, optional dependencies. One current
solution is to define an extra called, for example, ``recommended``, which
includes all non-essential but suggested dependencies. Users are then told to

> install the package using ``package[recommended]``, while those who prefer more

Reviewer 2: Possibly covered later (in which case, ignore), but why is the proposed approach better than something like `package[minimal]` or `package[default]`?

Author: @warsawnv - just to check, did the rest of the PEP answer this? This PEP in fact enables ``package[minimal]``, without this PEP it is impossible to make ``package[minimal]`` be a more minimal installation than ``package`` on its own. Does this make sense?

Reviewer: @astrofrog I do think we should add a few words along the lines "a solution will be detailed later on how to go about this problem".

We don't have to get into the specifics here, however mentioning that there should be a way to "deactivate" this behavior is a good idea IMHO.

Author: Sounds good, I have rephased this sentence to also not make it sound like 'one current solution' was the goal of this PEP, but rather a solution in the current status quo. I have also added a sentence to say that the goal of this PEP is to solve this.

control can simply use ``package``. However, in practice, many users are unaware
of the ``[recommended]`` syntax, placing the burden on them to know this for a
typical installation. Having a way to have recommended dependencies be installed
by default while providing a way for users to request a more minimal installation
would satisfy this use case.

Packages supporting multiple backends or frontends
--------------------------------------------------

Another common use case for using extras are to define different backends or
frontends, and dependencies that need to be installed for each backend or
frontend. A package might need at least one backend or frontend to be installed
in order to be functional, but may be flexible on which backend or frontend this
is. With current packaging standards, maintainers have the choice between either
making one of the backends or frontends be always required, or requiring users
to always specify extras, e.g. ``package[backend]``. Having a way to specify one
or more default backend or frontend and providing a way to override these

> defaults would provide a much better experience for users.

Reviewer 3: I read this yesterday and couldn't think of an actual package that does this. I just found one today; `viscm` needs to be installed with either extra `PyQt` or `Pyside`, but isn't functional on its own. Just in case you need a concrete example here.

Author: Thanks! Same with glue - I wasn't sure if I should single out specific packages in a PEP though?

Reviewer 3: I think it's okay (especially if you can give more than one illustration); I believe I've seen it done in other packaging-related PEPs.

Reviewer 2: > Just in case you need a concrete example here.

Concrete existing examples would help with the motivation.

Reviewer 4: An example where this is not currently done but could be useful in the future in a pretty high-impact area is with switching between BLAS/LAPACK implementations, for example in numpy/scipy. In those cases it is also not done yet due to additional considerations (ABI compat etc), but it may still be a useful example. Alternative FFT implementations (e.g. pyfftw) could be another useful example.

Reviewer: Agreed with Vyas & Barry.

@astrofrog can we add a few of the aforementioned examples ?

Author: I've added some concrete examples

Specification
=============

``Default-Extra`` Metadata Field
---------------------------------

A new metadata field, ``Default-Extra``, will be added to the `core package
metadata <https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata>`_.
This field allows package maintainers to define an extra that is
automatically included when a user installs the package without specifying any
extras explicitly::

    Default-Extra: recommended

If multiple default extras are needed, one ``Default-Extra:`` entry
should be provided for each one::

>     Default-Extra: backend1

Reviewer: No opinion here. Is it better for any reason to do:

```
Default-Extra: backend1
Default-Extra: backend2
Default-Extra: backend3
```

Instead of:
```
Default-Extra: backend1,backend2,backend3
# or
Default-Extras: backend1,backend2,backend3
```

It might be an implementation artifact of how METADATA works, is there a "reason" to go for one over the other ?

Reviewer 5: The former is a supported thing in the email header format, thus does not require additional parsing on the part of downstream consumers of the format.

Besides that, I don't think there's any no strong reason to have an invented mini-format (eg: do we strip whitespace around commas, are any other separators valid, etc) when the overall format supports the relevant functionality transparently.

Author: I originally had it as comma-separated values but changed it following @RonnyPfannschmidt's suggestion in https://github.com/astrofrog/peps/pull/1#discussion_r1775362470 - I don't have strong opinions

Reviewer 6: Comma separated is not the norm for metadata items. I'd stick with a multiple-use field.

Reviewer: Fair with me. Let's keep that way you can resolve this comment. Thanks for the pointers

    Default-Extra: backend2
    Default-Extra: backend3

Overriding default extras
-------------------------

The proposal here is simple - if extras are explicitly given in a dependency
specification, the default extras are not installed. For example, if a package
defines a ``extra1`` default extra as well as a non-default ``extra2``
extra, then if a user was to install the package with::

    pip install package

the ``extra1`` dependency would be included. If the user instead uses::

    pip install package[extra2]

then the ``extra1`` extras would not be installed.

If the same package is specified multiple times in a dependency tree, then if
any of these were specified without any explicit extras, the default extras

> should be installed even if other occurences of the package in the dependency

Reviewer 4: ```suggestion
should be installed even if other occurrences of the package in the dependency
```

tree have explicitly specified extras.

Note that ``package[]`` would continue to be equivalent to ``package`` and would
not be provided as a way to install without default extras (see the `Rejected

> Alternatives`_ section for the rationale).

Reviewer: I think before switching to the next section, we should add a subsection "solution to install without default"

Author: I added a new section 'Installing without default extras'

Usage scenarios
===============

The simple rule above regarding only installing default extras when no extras
are explicitly specified, combined with the introduction of the
``Default-Extra:`` keyword allows us to address several different use cases.
Below we take a look at the two specific use cases raised in the `Motivation`_
section as well as an additional case.

Recommended dependencies and minimal installations
--------------------------------------------------

First, we consider the case of packages that want users to get recommended
but not strictly required dependencies installed by default, while also
providing a way to only install the required dependencies.

In order to do this, a package maintainer would define an extras called
``recommended`` containing the recommended but not required dependencies, and
would choose to have this be included as a default extras. Exactly how this is
done will depend on the packaging tool, but one could imagine doing something

> as follows in a ``pyproject.toml`` file:

Reviewer 2: I think this PEP should take a stronger stance for proposing `pyproject.toml` syntax for default dependencies, especially in light of the recent acceptance of [PEP 735](https://peps.python.org/pep-0735/) which defines dependency groups.  In fact, perhaps something like:

```toml
[dependency-groups]
default-extras = [
    'foo',
    'bar',
]
```

could be possible syntax.

But even if not, I think some mention of this PEP's relationship to PEP 735 is needed.

Reviewer 5: > [dependency-groups] default-extras 

FWIW, I'm strongly opposed to having _this_ specific syntax here.

That said, I agree that this PEP should also be standardising the project table key here.

> But even if not, I think some mention of this PEP's relationship to PEP 735 is needed.

FWIW, I'm in the camp of folks who think that allowing mixing of dependency groups and extras is a recipe for user confusion.

Plus, as it stands, PEP 735 does not affect package metadata so (IMO) the only valid stance in this PEP is "there are no effects to consider".

Author: I've added a new section to the PEP, 'New key in ``[project]`` metadata table' (will push shortly)

PEP: 9999
Title: Default Extras for Python Software Packages
Author: Thomas Robitaille <thomas.robitaille@gmail.com>, Jonathan Dekhtiar <jonathan@dekhtiar.com>

> Sponsor: TBD

Reviewer 5: ```suggestion
Sponsor: Pradyun Gedam <pradyunsg@gmail.com>
```

Status: Draft
Type: Standards Track
Topic: Packaging
Created: 24-Sep-2024

Abstract
========

:pep:`508` specifies a mini-language for
declaring package dependencies. One feature of this language is the ability to
specify *extras*, which are optional components of a distribution that, when
used, install additional dependencies. This PEP proposes a mechanism to allow
one or more extras to be installed by default if none are provided explicitly.

Motivation
==========

Various use cases for default extras and possible solutions in this PEP were discussed

> extensively on `this DPO thread <https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/38>`__.

Reviewer 4: Is it worth listing other potential use cases without the same level of elaboration as the two below, or do you think that would simply muddle the proposal without adding any value?

Author: I think that the two use cases I mentioned actually cover most of the ones I saw in the various discussions - the second discussion on multiple front-ends or backends is broad enough to cover several cases that were mentioned such as having different computational backends, or different GUI frameworks and so on. I have reworded this to say:

_Various use cases for default extras and possible solutions in this PEP were discussed extensively on this DPO thread. These fall into two broad cases that that provide the motivation for the present PEP._

to not make it sound like I just cherry picked two use cases and ignored others. The two use cases here are a synthesis of the different cases I saw in the discussion. However if anyone would like to bring up another distinct use case, I'd be happy to add it!

categorized as extras. A balance must be struck between the needs of typical
users (who may prefer most features to be available 'by default') and users who
want minimal installations without large, optional dependencies. One solution
with existing Python packaging infrastructure is for package maintainers to
define an extra called, for example, ``recommended``, which
includes all non-essential but suggested dependencies. Users are then told to
install the package using ``package[recommended]``, while those who prefer more
control can simply use ``package``. However, in practice, many users are unaware
of the ``[recommended]`` syntax, placing the burden on them to know this for a
typical installation. Having a way to have recommended dependencies be installed
by default while providing a way for users to request a more minimal installation
would satisfy this use case. The present PEP will describe a solution for this use case.

Packages supporting multiple backends or frontends
--------------------------------------------------

Another common use case for using extras are to define different backends or

> frontends, and dependencies that need to be installed for each backend or

Reviewer 4: ```suggestion
Another common use case for using extras is to define different backends or
frontends and dependencies that need to be installed for each backend or
```

frontend. A package might need at least one backend or frontend to be installed
in order to be functional, but may be flexible on which backend or frontend this
is. Concrete examples of such frontends or backends include:

* The Qt frontend library, which requires either PySide or PyQt to be installed
* BLAS/LAPACK, which have different possible implementations (e.g. OpenBLAS, and MKL)
* FFT libraries, which also have different implementations (e.g. ``scipy.fft`` and pyFFTW)

With current packaging standards, maintainers have the choice between either
making one of the backends or frontends be always required, or requiring users
to always specify extras, e.g. ``package[backend]`` and therefore risking having
users have an unusable installation if they simply install ``package``. Having a

> way to specify one or more default backend or frontend and providing a way to

Reviewer 4: ```suggestion
With current packaging standards, maintainers have to either
require one of the backends or frontends, or require users
to always specify extras, e.g. ``package[backend]`` and therefore risk users
having an unusable installation if they simply install ``package``. Having a
way to specify one or more default backend or frontend and providing a way to
```

override these defaults would provide a much better experience for users, and
the approach described in this PEP will allow this.

Note that this PEP does not aim to address the issue of disallowing conflicting
or incompatible extras - for example if a package requires exactly one frontend
or backend package. There is currently no mechanism in Python packaging
infrastructure to disallow conflicting or incompatible extras to be installed,

> and this PEP does not change this.

Reviewer 4: ```suggestion
infrastructure to disallow conflicting or incompatible extras to be installed,
and this PEP does not change that.
```

Rationale
=========

A number of possible solutions have been extensively and vigorously discussed by
the community for several years, including in `this DPO thread
<https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/38>`__
as well as in numerous issues and pull requests. The solution that will be
presented below may not be everyone's preferred approach from a philosophical
point of view, however it appears to be the best compromise as it is the only
solution to a very real problem that would not break backward-compatibility of

> packaging infrastructure in any significant way. It is an opt-in solution which

Reviewer 4: I haven't been involved in the prior conversations so perhaps I didn't see the discussions that led to this acknowledgment, but naively it doesn't seem like it adds much here. If you're concerned about getting philosophical pushback from particular reviewers though feel free to keep it.

Author: Let's just say there has been some philosophical pushback :sweat_smile: - obviously I'd be happy to remove this if people object to including it, but until we get to the point where we have some kind of consensus, it might be safer to keep.

means that package maintainers can choose whether or not to use it, and it is
flexible enough to accommodate both of the major use cases described in
`Motivation`_.

Specification
=============

``Default-Extra`` Metadata Field
---------------------------------

A new metadata field, ``Default-Extra``, will be added to the `core package
metadata <https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata>`_.
This field allows package maintainers to define an extra that is
automatically included when a user installs the package without specifying any
extras explicitly::

    Default-Extra: recommended

If multiple default extras are needed, one ``Default-Extra:`` entry
should be provided for each one::

    Default-Extra: backend1
    Default-Extra: backend2
    Default-Extra: backend3

New key in ``[project]`` metadata table
---------------------------------------

A new key will be added to the ``[project]`` table in project metadata as
originally defined in :pep:`621` and now defined in the `PyPA specifications
<https://packaging.python.org/en/latest/specifications/>`_. This key will be named

> ``default-optional-depencencies`` with the following description:

Reviewer 4: ```suggestion
``default-optional-dependencies`` with the following description:
```

* `TOML <https://toml.io/>`_ type: Array of strings
* Corresponding core metadata field: ``Default-Extra``

> Each string in ``default-optional-depencencies`` should be the name of an extra

Reviewer 4: ```suggestion
Each string in ``default-optional-dependencies`` should be the name of an extra
```

defined in `optional-dependencies
<https://packaging.python.org/en/latest/specifications/pyproject-toml/#dependencies-optional-dependencies>`_,
and each extra in this array will be converted to a matching ``Default-Extra``
entry in the core package metadata. Examples of valid usage which would
produce the example ``Default-Extra`` entries presented in the previous section are:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "recommended"
    ]

and:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "backend1",
        "backend2",
        "backend3"
    ]

Overriding default extras
-------------------------

> The proposal here is simple - if extras are explicitly given in a dependency
> specification, the default extras are not installed. For example, if a package
> defines a ``extra1`` default extra as well as a non-default ``extra2``
> extra, then if a user was to install the package with::

Reviewer 5: ```suggestion
If extras are explicitly given in a dependency specification, the default extras are not installed. Otherwise, the default extras are used.

For example, if a package
defines a ``extra1`` default extra as well as a non-default ``extra2``
extra, then if a user was to install the package with::
```

A suggested rephrasing to avoid saying this is a proposal (in a subsection called specification) + to call out the behaviour separately from the examples.

options. To take a specific example, a package may need either PyQt or PySide to
be installed but will not work if none are provided, so a package maintainer may
therefore not want to provide an option to install the package without any
extras.

Backward Compatibility
======================

All package specification cases valid under :pep:`508` will remain valid.
Therefore, this proposal is fully backward-compatible with existing :pep:`508`
usage.

Once packages start defining default extras, those defaults will only be honored
with recent versions of packaging tools which implement this PEP, but those
packages will remain fully backward-compatible with older packaging tools - with
the only difference that the default extras will not be installed automatically
for older packaging tools.

The only conceptual backward-compatibility issue to consider is the fact that
this PEP changes extras to no longer be strictly additive, in that specifying
an extra such as ``minimal`` could result in fewer packages being installed.

Security Implications
=====================

There are no known security implications for this PEP.

How to teach this
=================

> The simple rule above regarding only installing default extras when no extras

Reviewer 5: Suggestion: Follow the guidance in https://justsimply.dev and removing the use of "just" / "simple" and similar words from this entire PEP.

Author: Done, I have checked for these and other similar words

are explicitly specified, combined with the introduction of the
``Default-Extra:`` keyword and ``default-optional-dependencies`` metadata key
allows us to address several different use cases. Below we take a look at the
two specific use cases raised in the `Motivation`_ section and how package
maintainers should be taught to address these.

Recommended dependencies and minimal installations
--------------------------------------------------

> First, we consider the case of packages that want users to get recommended

Reviewer 4: ```suggestion
First, we consider the case of packages that want recommended
```

but not strictly required dependencies installed by default, while also
providing a way to only install the required dependencies.

In order to do this, a package maintainer would define an extras called
``recommended`` containing the recommended but not required dependencies, and
would choose to have this be included as a default extras:

.. code-block:: toml

    [project]

>     default-optional-dependencies = [

Reviewer 5: I'm not a huge fan of this name, since this is listing the keys under `optional-dependencies` rather than PEP 508 strings. OTOH, I think we can pick the bike shed color in the d.p.o. thread. :)

Author: Yes I see what you mean - I have no strong opinion here and happy to update it if someone comes up with a better suggestion.

        "recommended"
    ]

    [project.optional-dependencies]
    recommended = [
        "package1",
        "package2"
    ]

In this specific case, a package maintainer may want to allow users to also
install the package without the recommended dependencies, in which case they
could define an empty extras:

.. code-block:: toml

    [project.optional-dependencies]
    minimal = []
    recommended = [
        "package1",
        "package2"
    ]

This would then allow users to install ``package[minimal]`` which, since

> there would be an extras explicitly specified, would mean the default extras

Reviewer 4: ```suggestion
there would be an extra explicitly specified, would mean the default extras
```

as required dependencies rather than using the default extras mechanism.

Supporting minimal installations while not always removing default extras
-------------------------------------------------------------------------

An additional case we consider here is where a package maintainer wants to support
minimal installations without any extras, but also wants to support having users
specify additional extras without removing the default one. Essentially, they
would want:

* ``package[minimal]`` to give an installation without any extras
* ``package`` to install recommended dependencies (in a ``recommended`` extras)
* ``package[additional]`` to install both recommended and additional dependencies (in an ``additional`` extras)

This could be achieved with e.g:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "recommended"
    ]

    [project.optional-dependencies]
    minimal = []
    recommended = [
        "package1",
        "package2"
    ]
    additional = [
        "package[recommended]",
        "package3"
    ]

The ability for a package to reference itself in the extras is supported by
existing Python packaging tools.

> Teaching to package users
> -------------------------

Reviewer 5: ```suggestion
Teaching package authors
------------------------
```

2 things: "teaching to" does not read as valid English grammar to me & I think it's worth being specific about what the specific user we care about (authors vs consumers, i.e. if it's the people uploading packages or people downloading packages).

(``-``) before the extra name. The proposed syntax update would have been as follows::

    extras_list   = (-)?identifier (wsp* ',' wsp* (-)?identifier)*

Valid examples of this new syntax would have included, e.g.:

* ``package[-recommended]``
* ``package[-backend1, backend2]``
* ``package[pdf, -svg]``

However, there are two main issues with this approach:

* One would need to define a number of rules for how to interpret corner cases
  such as if an extras and its negated version were both present in the same
  dependency specification (e.g. ``package[pdf, -pdf]``) or if a dependency
  tree included both ``package[pdf]`` and ``package[-pdf]``, and the rules would
  not be intuitive to users.

* More critically, this would introduce new syntax into dependency specification,
  which means that if any package defined a dependency using the new syntax, it
  and any other package depending on it would no longer be installable by existing
  packaging tools, so this would be a major backward compatibility break.

For these reasons, this alternative was not included in the final proposal.

Adding a special entry in ``extras_require``

> --------------------------------------------

Reviewer 5: I think this section should cover why "default" is not a good name for the special entry here.

Author: I have made a tweak to this section to change:

_In addition, no other string can be used as a special
string since all strings that would be a backward-compatible valid extras name
may already be used in existing packages._

to

_In addition, no other string (such as ``'default'``) can be used as a special
string since all strings that would be a backward-compatible valid extras name
may already be used in existing packages._

Hopefully this addresses your comment?

Reviewer 5: (I can't click the resolve button) Yes. :)
